### PR TITLE
Dist2 initial improvements

### PR DESCRIPTION
**Pull request by brian-kelley**
- Fix distance-2 algorithms not checking for columns being in-bounds  (meaning it was not usable for MueLu aggregation on distributed problems)
- Remove the serial distance-2 algorithm from the distance-1 handle
- Implement a new net-based serial algorithm (asymptotically better than the neighbors-of-neighbors nested loop used in vertex-based). In practice, is faster than ColPack by >10x and both the previous serial algorithm and Zoltan by >18x.
- Implement a new parallel net-based algorithm (``COLORING_D2_NB_BIT``). It gives a 2x speedup over VB and VB_BIT on GPU, but can still be improved.
- Improve the user interface to the D1 and D2 coloring performance tests

Also, @srajama1 and I just found out about a serious bug in all D2 algorithms where D1 conflicts ignored (the algorithms assume that diagonal edges exist, but this is wrong). The D2 coloring verification has the same bug, so it was never caught in testing. This PR is still a step in the right direction (doesn't introduce bugs) so I will just submit it as-is and work on the next fix on top of these changes.

RIDE spot check:
#######################################################
PASSED TESTS
#######################################################
cuda-10.1.105-Cuda_OpenMP-release build_time=619 run_time=416
cuda-10.1.105-Cuda_Serial-release build_time=498 run_time=518
cuda-9.2.88-Cuda_OpenMP-release build_time=550 run_time=418
cuda-9.2.88-Cuda_Serial-release build_time=521 run_time=520
gcc-6.4.0-OpenMP_Serial-release build_time=215 run_time=393
gcc-7.2.0-OpenMP-release build_time=122 run_time=128
gcc-7.2.0-OpenMP_Serial-release build_time=194 run_time=360
gcc-7.2.0-Serial-release build_time=120 run_time=226
ibm-16.1.0-Serial-release build_time=490 run_time=392

Bowman spot check:
#######################################################
PASSED TESTS
#######################################################
intel-16.4.258-Pthread-release build_time=768 run_time=1033
intel-16.4.258-Pthread_Serial-release build_time=1112 run_time=1910
intel-16.4.258-Serial-release build_time=723 run_time=820
intel-17.2.174-OpenMP-release build_time=958 run_time=560
intel-17.2.174-OpenMP_Serial-release build_time=1364 run_time=1456
intel-17.2.174-Pthread-release build_time=870 run_time=896
intel-17.2.174-Pthread_Serial-release build_time=1298 run_time=1781
intel-17.2.174-Serial-release build_time=827 run_time=883

### CODE REVIEW

**Comment by srajama1** [comment]
Let me understand this, we still have the d2 bug when diagonal entries are not present. Otherwise, things are ok ?

**Comment by brian-kelley** [comment]
@srajama1 Correct.

**Comment by srajama1** [comment]
Ok, reviewing it now. Can we file two issues, one for these changes. one for the missing diagonal giving wrong results ?

**Comment by brian-kelley** [comment]
Created those issues: #625, and #624.

File: src/graph/impl/KokkosGraph_Distance2Color_impl.hpp [hunk 1]
```diff
@@ -405,10 +409,530 @@ class GraphColorDistance2
...
+      }
+
+      color_type colorBase;
+      forbidden_view forbidden;         //forbidden color bitset for columns
+      color_view_type colors;
+      const_lno_row_view_type Crowmap;  //C <-> V graph (row c is the vertices incident to c)
+      const_lno_nnz_view_t    Ccolinds;
+      const nnz_lno_type numVerts;
+    };
+
```
Comment (srajama1) [inline src/graph/impl/KokkosGraph_Distance2Color_impl.hpp]: It would be nice to write this algorithm up in latex.

**Comment by srajama1** [comment]
Thanks, @brian-kelley ! The issues will help @ndellingwood track these for the release. We also need to update the documentation.

File: src/graph/KokkosGraph_Distance2ColorHandle.hpp [hunk 2]
```diff
@@ -205,60 +206,70 @@ class GraphColorDistance2Handle
...
+            found = true;
 #ifdef VERBOSE
-            std::cout << "Qthread Execution Space, Default Algorithm: COLORING_VB" << std::endl;
+            std::cout << "Qthread Execution Space, Default Algorithm: COLORING_D2_NB_BIT" << std::endl;
 #endif
         }
 #endif
+        //Since this logic is based on checking every exec space, detect when a new one needs to be supported
+        if(!found)
+          throw std::logic_error("D2 coloring: default algorithm hasn't been chosen for th current execution space");
```
[the remark this answers is not in the document]
Comment (brian-kelley) [reply src/graph/KokkosGraph_Distance2ColorHandle.hpp]: Thanks, fixed

**Pull request merged**